onnxruntime/core/session/inference_session.cc

```diff
@@ -161,12 +162,12 @@ static Status FinalizeSessionOptions(const SessionOptions& user_provided_session
 
 void InferenceSession::ConstructorCommon(const SessionOptions& session_options,
                                          logging::LoggingManager* logging_manager) {
-  auto status = FinalizeSessionOptions(session_options, model_proto_.get(), session_options_);
+  auto status = FinalizeSessionOptions(session_options, model_proto_, model_proto_parsed_, session_options_);
   ORT_ENFORCE(status.IsOK(), "Could not finalize session options while constructing the inference session. Error Message: ",
               status.ErrorMessage());
 
-  graph_transformation_mgr_ = onnxruntime::make_unique<GraphTransformerManager>(
-      session_options_.max_num_graph_transformation_steps);
+  graph_transformation_mgr_.Init(session_options_.max_num_graph_transformation_steps);
+
```

Comment:
This seems like the wrong balance. In general you don't want to have classes with separate init methods. 
[https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-complete](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-complete)

Why can't we use a default for steps and allow it to be overridden here?

Reply:
I understand the concern. We could create an instance of GraphTransformerManager with default steps and allow it be overridden if the value of steps from the *finalized* session_options is different from the default. But to allow arbitrary changes to steps in the same instance of GraphTransformerManager seems a little dangerous to me. Hence, I went with an approach so as to allow it getting set once only...